internal/batch/batch_processor.go

```diff
@@ -101,103 +130,249 @@ func (bw *batchWork) estimateSize() int64 {
 	return sizeEstimate
 }
 
-// The assemblyLoop accepts work into the pipe as quickly as possible.
-// It dispatches work asynchronously to the persistenceLoop, which is responsible for
-// calling back each piece of work once persisted into a batch
-// (doesn't wait until that batch is sealed/dispatched).
-// The assemblyLoop seals batches when they are full, or timeout.
+func (bp *batchProcessor) status() *ProcessorStatus {
+	bp.statusMux.Lock()
+	defer bp.statusMux.Unlock()
+	return &ProcessorStatus{
+		Dispatcher: bp.conf.dispatcherName,
+		Name:       bp.conf.name,
+		Status:     bp.flushStatus, // copy
+	}
+}
+
+func (bp *batchProcessor) newAssembly(initalWork ...*batchWork) {
+	bp.assemblyID = fftypes.NewUUID()
+	bp.assemblyQueue = append([]*batchWork{}, initalWork...)
+	bp.assemblyQueueBytes = batchSizeEstimateBase
+}
+
+// addWork adds the work to the assemblyQueue, and calculates if we have overflowed with this work.
+// We check for duplicates, and add the work in sequence order.
+// This helps in the case for parallel REST APIs all committing to the DB at a similar time.
+// With a sufficient batch size and batch timeout, the batch will still dispatch the messages
+// in DB sequence order (although this is not guaranteed).
+func (bp *batchProcessor) addWork(newWork *batchWork) (full, overflow bool) {
+	newQueue := make([]*batchWork, 0, len(bp.assemblyQueue)+1)
+	added := false
+	skip := false
+	// Check it's not in the recently flushed lish
+	for _, flushedSequence := range bp.flushedSequences {
+		if newWork.msg.Sequence == flushedSequence {
+			log.L(bp.ctx).Debugf("Ignoring add of recently flushed message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			skip = true
+			break
+		}
+	}
+	// Build the new sorted work list, checking there for duplicates too
```

Comment:
If `skip` is true by this point, we're going to do a bunch of looping to make a copy of `bp.assemblyQueue` and then replace it with a copy of itself. Feels like if `skip` is set at any point, we can just abandon `newQueue` altogether and keep the existing queue intact, right?